### PR TITLE
Added support for optional scopes in interactive and silent requests

### PR DESCRIPTION
**Pull request by oldalton**

### CODE REVIEW

File: MSAL/src/public/MSALError.h [hunk 1]
```diff
@@ -60,6 +79,26 @@ extern NSString *MSALCorrelationIDKey;
  */
 extern NSString *MSALHTTPResponseCodeKey;
 
+/*!
+ List of scopes that were requested from MSAL, but not granted in the response.
+
+ Apps normally request certain permissions that their app needs to function.
+ However, sometimes STS grants only a subset of those permissions.
```
Comment (unpluggedk) [inline MSAL/src/public/MSALError.h]: nit: I don't think this adds much, first line covers this.

File: MSAL/src/public/MSALError.h [hunk 2]
```diff
@@ -60,6 +79,26 @@ extern NSString *MSALCorrelationIDKey;
...
+
+ This can happen in multiple cases:
+
+    * Requested scope is not supported
+    * Requested scope is not Recognized (According to OIDC, any scope values used that are not understood by an implementation SHOULD be ignored.)
+    * Requested scope is not supported for a particular account (Organizational scopes when it is a consumer account)
```
Comment (unpluggedk) [inline MSAL/src/public/MSALError.h]: nit: Organization scopes asked from a consumer account ?

File: MSAL/src/public/MSALError.h [hunk 3]
```diff
@@ -201,7 +242,7 @@ typedef NS_ENUM(NSInteger, MSALErrorCode)
     
     /*!
      Response was received in a network call, but the response body was invalid.
-     R
+
```
Comment (unpluggedk) [inline MSAL/src/public/MSALError.h]: what was this R??

File: MSAL/src/requests/MSALBaseRequest.m [hunk 4]
```diff
@@ -195,52 +195,111 @@ - (void)acquireToken:(nonnull MSALCompletionBlock)completionBlock
...
-        {
-            completionBlock(nil, [MSALErrorConverter MSALErrorFromMSIDError:localError]);
-            return;
-        }
-        
-        if (_parameters.account.homeAccountId.identifier != nil &&
-            ![_parameters.account.homeAccountId.identifier isEqualToString:tokenResponse.clientInfo.accountIdentifier])
+
+        NSError *verificationError = nil;
+        if (![self verifyTokenResponse:tokenResponse error:&verificationError])
```
Comment (unpluggedk) [inline MSAL/src/requests/MSALBaseRequest.m]: 👍

File: MSAL/src/requests/MSALBaseRequest.m [hunk 5]
```diff
@@ -195,52 +195,111 @@ - (void)acquireToken:(nonnull MSALCompletionBlock)completionBlock
...
+    MSIDConfiguration *configuration = _parameters.msidConfiguration;
+
+    if (![configuration.scopes isSubsetOfOrderedSet:grantedScopes])
+    {
+        if (error)
+        {
+            NSMutableDictionary *additionalUserInfo = [NSMutableDictionary new];
+            additionalUserInfo[MSALGrantedScopesKey] = [grantedScopes array];
+
+            NSMutableOrderedSet *requestedScopeSet = [configuration.scopes mutableCopy];
```
Comment (unpluggedk) [inline MSAL/src/requests/MSALBaseRequest.m]: nit: declinedScopeSet = [configuration.scopes mutableCopy] minusOrderedSet... ?
That way, next line would read easier.

File: MSAL/src/public/MSALError.h (hunk 1 above)
Comment (oldalton) [reply MSAL/src/public/MSALError.h]: done

File: MSAL/src/public/MSALError.h (hunk 2 above)
Comment (oldalton) [reply MSAL/src/public/MSALError.h]: yes

File: MSAL/src/public/MSALError.h (hunk 3 above)
Comment (oldalton) [reply MSAL/src/public/MSALError.h]: dunno

File: MSAL/src/requests/MSALBaseRequest.m (hunk 5 above)
Comment (oldalton) [reply MSAL/src/requests/MSALBaseRequest.m]: agree, fixed

**Pull request merged**